### PR TITLE
[GLUTEN-3475] Add spark35 scala213

### PR DESCRIPTION
**Pull request by holdenk**
## What changes were proposed in this pull request?

(Please fill in changes proposed in this fix)

Fixes: #3475

## How was this patch tested?

In progress, existing tests and compiling

See https://www.youtube.com/watch?v=UbB7v2f7A6k

### CODE REVIEW

File: dev/buildbundle-veloxbe.sh [hunk 1]
```diff
@@ -7,3 +7,6 @@ cd $GLUTEN_DIR
 mvn clean package -Pbackends-velox -Prss -Pspark-3.2 -DskipTests
 mvn clean package -Pbackends-velox -Prss -Pspark-3.3 -DskipTests
 mvn clean package -Pbackends-velox -Prss -Pspark-3.4 -DskipTests
+mvn clean package -Pbackends-velox -Prss -Pspark-3.5 -DskipTests
+# No celbron Scala 2.13 support yet.
```
Comment (LuciferYang) [inline dev/buildbundle-veloxbe.sh:11]: Typo: celeborn

File: backends-clickhouse/pom.xml [hunk 2]
```diff
@@ -79,7 +79,7 @@
     <dependency>
       <groupId>org.scalacheck</groupId>
       <artifactId>scalacheck_${scala.binary.version}</artifactId>
-      <version>1.13.5</version>
+      <version>1.14.3</version>
```
Comment (LuciferYang) [inline backends-clickhouse/pom.xml:82]: https://github.com/oap-project/gluten/blob/d843bd93b460d6988c5d654d57978d096b0a835a/backends-clickhouse/pom.xml#L106-L117

There are some related dependencies that use hard coding `_ 2.12` in this `pom.xml`, they should also use `${scala.binary.version}`, I suspect there are other similar cases

File: gluten-core/src/main/scala/io/glutenproject/GlutenPlugin.scala [hunk 3]
```diff
@@ -190,7 +190,8 @@ private[glutenproject] class GlutenDriverPlugin extends DriverPlugin with Loggin
 
 private[glutenproject] class GlutenExecutorPlugin extends ExecutorPlugin {
   private var executorEndpoint: GlutenExecutorEndpoint = _
-  private val taskListeners: Seq[TaskListener] = Array(TaskResources)
+  private val taskListeners: Seq[TaskListener] =
+    (immutable.ArraySeq.unsafeWrapArray(Array(TaskResources)))
```
Comment (LuciferYang) [inline gluten-core/src/main/scala/io/glutenproject/GlutenPlugin.scala:194]: `immutable.ArraySeq.unsafeWrapArray` is a Scala 2.13 only api, if this performance optimization is necessary, this file may need to be split into Scala 2.12 and Scala 2.13 versions

File: gluten-core/src/main/scala/io/glutenproject/backendsapi/BackendsApiManager.scala [hunk 4]
```diff
@@ -20,7 +20,7 @@ import io.glutenproject.GlutenPlugin
 
 import java.util.ServiceLoader
 
-import scala.collection.JavaConverters
+import scala.jdk.CollectionConverters._
```
Comment (LuciferYang) [inline gluten-core/src/main/scala/io/glutenproject/backendsapi/BackendsApiManager.scala:23]: IIRC, `scala.jdk.CollectionConverters` also Scala 2.13 only

**Review by LuciferYang — COMMENTED** [review]
Seems that some code has not been uploaded yet? I haven't been able to compile successfully locally yet

File: gluten-core/src/main/scala/io/glutenproject/backendsapi/BackendsApiManager.scala (hunk 4 above)
Comment (holdenk) [reply gluten-core/src/main/scala/io/glutenproject/backendsapi/BackendsApiManager.scala:23]: Ah good point, looks like there is a backwards compact lib ( https://stackoverflow.com/questions/59197883/version-agnostic-way-to-convert-from-java-to-scala-collections-and-back ) I'll try for Scala 2.12 support

**Comment by holdenk** [comment]
Still WIP but made some more progress

**Comment by holdenk** [comment]
@ LuciferYang  any chance I can get read access to Jenkins?

**Comment by holdenk** [comment]
@LuciferYang  any chance I can get read access to Jenkins?

**Comment by LuciferYang** [comment]
> @LuciferYang any chance I can get read access to Jenkins?

Sorry, I'm not familiar with this matter.

cc @zhouyuan

**Comment by holdenk** [comment]
Hey y'all, just following up I'm happy to update the PR on top of main if someone can get me access to CI and has the cycles to review, otherwise I'll just leave it be.

**Comment by JkSelf** [comment]
@holdenk   Thanks for your great work. I have approved the CI running. The changes in this PR are too extensive, making it difficult to review. I noticed that the Scala version in Spark 3.5 is still 2.12 [here](https://github.com/apache/spark/blob/v3.5.0/pom.xml#L175). So, can we enable only spark 3.5 in this PR and create a separate PR to upgrade the Scala version to 2.13 if necessary? Also, could you please help list some of the main modifications in the PR description to facilitate our review?

**Comment by holdenk** [comment]
@JkSelf thanks! Can I get a log in so I can see the CI results / logs?

As far as splitting out Scala 2.13 / Spark 3.5 changes certainly possible but given everything so far I'd want to know there was someone with merge permissions with the time and interest to do a review.

**Comment by JkSelf** [comment]
@holdenk You can click on the `Details `to view the CI log. I will follow up with the review later. Thanks.
image

**Comment by holdenk** [comment]
So the Jenkins instance requires a login, without that I'm not sure how to fully debug the issues which come up when I fix the GHA issues.

**Comment by JkSelf** [comment]
> So the Jenkins instance requires a login, without that I'm not sure how to fully debug the issues which come up when I fix the GHA issues.

@zhztheplayer Do you have any input? Thanks.

**Comment by zhztheplayer** [comment]
The Jenkins CI is only for ClickHouse backend, username/password: gluten/hN2xX3uQ4m , see [ref](https://github.com/oap-project/gluten/blob/main/docs/get-started/ClickHouse.md).

**Comment by holdenk** [comment]
Awesome, thanks for the login.

**Comment by JkSelf** [comment]
@holdenk It appears that the Velox backends jobs have also failed. Please ensure that all Velox and ClickHouse backends pass successfully.  Thanks.

**Comment by holdenk** [comment]
So if you're down to review the change sure I'll make a Scala 2.12 PR once I've got some time (maybe later this week or next), but unless there's a reviewer available I'm not working on this anymore.

**Comment by JkSelf** [comment]
Yes, Can we merge the Spark 3.5 PR first and then proceed with the Scala 2.13 PR?

**Comment by holdenk** [comment]
Rocking, I'll try to put together a Scala 2.12 Spark 3.5 PR for you to review for the first week of January :)

**Pull request closed without merging**